orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -38,18 +39,46 @@ def _ortvalue_from_torch_tensor(torch_tensor):
     return C.OrtValue.from_dlpack(to_dlpack(torch_tensor), is_bool_tensor)
 
 
-def _torch_tensor_from_dl_pack(dlpack, ortvalue, device):
-    torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack)
-    return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
+def _ortvalues_to_torch_tensor(ortvalues, device):
+    if len(ortvalues) == 0:
+        return tuple()
+    if hasattr(ortvalues, 'to_dlpack'):
+        if all(ortvalue.proto_type() != 9 for ortvalue in ortvalues):
+            if 'ort' == device.type and hasattr(C, 'ort_from_dlpack'):
```

Comment:
I suggest not to use IF clause to return different torch tensors when C has ort_from_dlpack or not. I understand that you make it this way because you still want to add some tests related to 'ort' device, but `C.ort_from_dlpack` and torch's _from_dlpack return different torch.Tensors. `C.ort_from_dlpack` returns a torch.Tensor that use `ORTTensorImpl` as the final implementation, which is written by ourselves, but torch's _from_dlpack returns a normal torch.Tensor implemented by torch. If the device is 'ort', returning a normal torch.Tensor (using torch's _from_dlpack) will cause failure when it goes to somewhere else in the backend at the end. So actually this IF is hiding some problem (why C doesn't have ort_from_dlpack when device is 'ort') here and may bring trouble to debugging. Assertion (that C must have ort_from_dlpack) should be the correct behavior here. (The original code calls `C.ort_from_dlpack` directly, and if C doesn't have ort_from_dlpack, Python will report the problem. It's better we don't introduce any new logic compared to the old code.) @souptc who is the developer of ort_eager mode may want to add more comment. Same comment applies to below code from line 54 to 57.